### PR TITLE
Add category

### PR DESCRIPTION
**Pull request by HuiminFrontend**

### CODE REVIEW

File: src/components/Categories/Categories.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import {useEffect, useState} from 'react';
+import styled from 'styled-components';
+
+export default function Categories({menus, handleCategoryClick, curClickedCategory}) {
+  const [categories, setCategories] = useState([]);
+
+  useEffect(() => {
+    /* Get categories form the given menus */
+    function getCategories(menus) {
+      const categories = menus.reduce(
```
Comment (stefankorder) [inline src/components/Categories/Categories.js:10]: I think reduce is not the right method for this case. Maybe there would be a other solution? What is your intention for this code?

File: src/components/Categories/Categories.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      );
+      return categories;
+    }
+
+    setCategories(getCategories(menus));
+
+    /* Sort categories in this order */
+    const sortBy = ['All', 'Starters', 'Mains', 'Pastries', 'Drinks'];
+    const categorySort = ({data, sortBy, sortfield}) => {
+      const sortByObj = sortBy.reduce(
```
Comment (stefankorder) [inline src/components/Categories/Categories.js:27]: Same as above.

File: src/components/Categories/Categories.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+          ...obj,
+          [item]: index,
+        }),
+        {}
+      );
+      return data.sort((a, b) => sortByObj[a[sortfield]] - sortByObj[b[sortfield]]);
+    };
+
+    categorySort({data: menus, sortBy, sortfield: 'category'});
+    /* eslint-disable-next-line */
```
Comment (stefankorder) [inline src/components/Categories/Categories.js:38]: Please delete this eslint-disable. Try to fix the error instead.

File: src/components/ShowMenus/ShowMenus.js [hunk 4]
```diff
@@ -1,15 +1,23 @@
...
-  const [menus, setmenus] = useState(menusdata);
+  useEffect(() => {
+    if (curClickedCategory === 'All') setRenderMenus(menus);
+    else
+      setRenderMenus(
+        menus.filter(menu => {
+          return curClickedCategory === menu.category;
+        })
+      );
+    /* eslint-disable-next-line */
```
Comment (stefankorder) [inline src/components/ShowMenus/ShowMenus.js:15]: Please delete this eslint-disable. Try to fix the error instead. Maybe useEffect is not the best solution for this case.

File: src/components/Categories/Categories.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
+import {useEffect, useState} from 'react';
+import styled from 'styled-components';
+
+export default function Categories({menus, handleCategoryClick, curClickedCategory}) {
+  const [categories, setCategories] = useState([]);
+
+  useEffect(() => {
+    /* Get categories form the given menus */
+    function getCategories(menus) {
```
Comment (stefankorder) [inline src/components/Categories/Categories.js:9]: You should test such complex functions.

**Review by stefankorder — CHANGES_REQUESTED** [review]
Take a look at my comments.

File: src/components/Categories/Categories.js (hunk 1 above)
Comment (HuiminFrontend) [reply src/components/Categories/Categories.js:10]: I want to find out what categories I have from the menus.

File: src/components/Categories/Categories.js (hunk 2 above)
Comment (HuiminFrontend) [reply src/components/Categories/Categories.js:27]: I can do render without this function, but the menus will be rendered in different order, like Mains, Starters.....  So this is the solution to the problem.

File: src/components/Categories/Categories.js (hunk 3 above)
Comment (HuiminFrontend) [reply src/components/Categories/Categories.js:38]: Yes, maybe you can help me out, useEffect need a missing dependency: menus, but if I add it to dependency array, it's also wrong! I receive another error. Add also I've found out, in some website, it's recommended to commit this out.

File: src/components/ShowMenus/ShowMenus.js (hunk 4 above)
Comment (HuiminFrontend) [reply src/components/ShowMenus/ShowMenus.js:15]:
```
  <MenusContainer>
          {menus.map(menu => {
          return curClickedCategory === menu.category ? (
           showMenu(menu)
          ) : curClickedCategory === 'All' ? (
           showMenu(menu)
          ) : (
           <></>
          );
       })}
    </MenusContainer>
```
At first I have a conditional render like this, the problem, it gives a error:  no unique key.  This means I have to give this thing <></>  a unique key, but to give this a key, it has to be div or some real html element, empty divs effect the layout.
Do you have any better idea?

File: src/components/Categories/Categories.js (hunk 5 above)
Comment (HuiminFrontend) [reply src/components/Categories/Categories.js:9]: Of course, my app works good, I test it!

File: src/components/Categories/Categories.js (hunk 3 above)
Comment (HuiminFrontend) [reply src/components/Categories/Categories.js:38]: As I can remenber Jerry has once also said we can commit it out. It should be a common problem, what do the others do under this situation?

File: src/components/ShowMenus/ShowMenus.js (hunk 4 above)
Comment (stefankorder) [reply src/components/ShowMenus/ShowMenus.js:15]: You need just the first else case. Either you have chosen a category or all will be displayed.

File: src/components/Categories/Categories.js (hunk 5 above)
Comment (stefankorder) [reply src/components/Categories/Categories.js:9]: But you dont test such functions with junit tests.

File: src/components/Categories/Categories.js (hunk 3 above)
Comment (stefankorder) [reply src/components/Categories/Categories.js:38]: Okay Gregor another headcoach told me it's not recommend to disable eslint in this case.

File: src/components/Categories/Categories.js (hunk 1 above)
Comment (stefankorder) [reply src/components/Categories/Categories.js:10]: function getAlternativeCategorie(menus) {
  const categories = ["All"]
  menus.forEach(menu => {
    if (!categories.includes(menu.category)) {
      categories.push(menu.category)
    }
  })
  return categories
}

File: src/components/ShowMenus/ShowMenus.js (hunk 4 above)
Comment (HuiminFrontend) [reply src/components/ShowMenus/ShowMenus.js:15]: > 

The problem is, only in map function I can know if a menu should be rendered or not, and map function should be always return something, even if I don't want a menu to be rendered, I have to return something, that means <></>, if I return this,  I need a unique key, to give a unique key, this thing have to be divs.

Comment (Huimin66) [reply src/components/ShowMenus/ShowMenus.js:15]: > You need just the first else case. Either you have chosen a category or all will be displayed.

Another idea,  I can do it when onClick happens, I'll let you know when I finish.

File: src/components/Categories/Categories.js (hunk 5 above)
Comment (HuiminFrontend) [reply src/components/Categories/Categories.js:9]: > 
Ok, I see, I'll do it. Thank you for tips, I didn't realize that pure function should also be tested.

File: src/components/Categories/Categories.js (hunk 3 above)
Comment (HuiminFrontend) [reply src/components/Categories/Categories.js:38]: > 

Jerry suggests to post this in question channel. Thank you anyway to bring this up, I also want to get a clear question about this.

File: src/components/ShowMenus/ShowMenus.js (hunk 4 above)
Comment (stefankorder) [reply src/components/ShowMenus/ShowMenus.js:15]: Is there a case you dont want to render anything?

Comment (HuiminFrontend) [reply src/components/ShowMenus/ShowMenus.js:15]: Hi Stefan, I just used filter and then map to solve the problem, it works good, less codes. Thank you very much!

File: src/components/Categories/Categories.js (hunk 1 above)
Comment (stefankorder) [reply src/components/Categories/Categories.js:10]: const sortBy = ["All", "Starters", "Mains", "Pastries", "Drinks"];

function getCategoriesAndSort(menus, sortBy) {
  const categories = ["All"];
  menus.forEach((menu) => {
    if (!categories.includes(menu.category)) {
      categories.push(menu.category);
    }
  });
  
  let sortByObj = {};
  sortBy.forEach((menuPoint, index) => {
    sortByObj = { ...sortByObj, [menuPoint]: index };
  });

  const sortedCategories = categories.sort(
    (a, b) => sortByObj[a] - sortByObj[b]
  );

  return sortedCategories;
}

File: src/components/Categories/Categories.js (hunk 2 above)
Comment (stefankorder) [reply src/components/Categories/Categories.js:27]: see my solution above

File: src/components/Categories/Categories.js (hunk 5 above)
Comment (stefankorder) [reply src/components/Categories/Categories.js:9]: Depends on the complexity.

File: src/components/Categories/Categories.js (hunk 2 above)
Comment (HuiminFrontend) [reply src/components/Categories/Categories.js:27]: Thank you, works good in my app.

**Pull request merged**